android/src/main/kotlin/io/chirp/chirpsdk/ChirpsdkPlugin.kt

```diff
@@ -49,127 +50,170 @@ class ChirpsdkPlugin(val activity: Activity) : MethodCallHandler {
+       * The payload argument contains the payload data that was sent.
+       */
+      activity.runOnUiThread {
+        sentStreamHandler.send(payload, channel)
       }
     }
-    else if (call.method == "stop") {
-      val error: ChirpError = chirpConnect.stop()
-      if (error.code > 0) {
-        errorStreamHandler.send(error.code, error.message)
+    chirpSDK.onReceiving { channel: Int ->
+      /**
+       * onReceiving is called when a receive event begins.
+       * No data has yet been received.
+       */
+      activity.runOnUiThread {
+        receivingStreamHandler.send(channel)
       }
     }
-    else if (call.method == "send") {
-      val payload: ByteArray =  call.arguments as ByteArray
-      val error: ChirpError = chirpConnect.send(payload)
-      if (error.code > 0) {
-        errorStreamHandler.send(error.code, error.message)
+    chirpSDK.onReceived { payload: ByteArray?, channel: Int ->
+      /**
+       * onReceived is called when a receive event has completed.
+       * If the payload was decoded successfully, it is passed in payload.
+       * Otherwise, payload is null.
+       */
+      activity.runOnUiThread {
+        receivedStreamHandler.send(payload, channel)
       }
     }
-    else if (call.method == "sendRandom") {
-      val payload: ByteArray = chirpConnect.randomPayload(0)
-      val error: ChirpError = chirpConnect.send(payload)
-      if (error.code > 0) {
-        errorStreamHandler.send(error.code, error.message)
+    chirpSDK.onStateChanged { oldState: ChirpSDKState, newState: ChirpSDKState ->
+      /**
+       * onStateChanged is called when the SDK changes state.
+       */
+      activity.runOnUiThread {
+        stateStreamHandler.send(oldState.code, newState.code)
       }
     }
-    // else if (call.method == "isValidPayload") {
-    //   val payload: ByteArray =  call.arguments as ByteArray
-    //   val error: ChirpError = chirpConnect.send(payload)
-    //   if (error.code > 0) {
-    //     errorStreamHandler.send(error.code, error.message)
-    //   }
-    // }
-    else if (call.method == "getState") {
-      result.success(chirpConnect.getState().code)
-    }
-    else if (call.method == "maxPayloadLength") {
-      result.success(chirpConnect.maxPayloadLength())
+    chirpSDK.onSystemVolumeChanged { oldVolume: Float, newVolume: Float ->
+      /**
+       * onSystemVolumeChanged is called when the system volume is changed.
+       */
     }
-    else if (call.method == "channelCount") {
-      result.success(chirpConnect.getChannelCount())
-    }
-    else if (call.method == "transmissionChannel") {
-      result.success(chirpConnect.getTransmissionChannel())
+  }
+
+  private fun setConfig(call: MethodCall, result: Result) {
+    if (!isInitialised(call, result)) return
+    val config: String = call.arguments as String
+    val error: ChirpError = chirpSDK.setConfig(config)
+
+    //TODO: brainstorm if this is the best approach, maybe we should just report error here?
```

Comment:
As you mentioned, the Flutter program calls these methods inside try/catch statements anyway, so I think we should use the suggested error handling methods as detailed here - https://flutter.dev/docs/development/platform-integration/platform-channels

```
val error: ChirpError = chirpSDK.setConfig(config)
if error.code == CHIRP_SDK_OK {
   result.success()
} else {
  result.error("ChirpSDK Error - {error.code}", errorCodeToString(error.code), null)
}

Reply:
I hate this idea tbh but you are right, we have to follow the documentation